pdns/dnsdist.cc

```diff
@@ -721,6 +754,47 @@ shared_ptr<DownstreamState> whashed(const NumberedServerVector& servers, const D
   return valrandom(dq->qname->hash(g_hashperturb), servers, dq);
 }
 
+shared_ptr<DownstreamState> chashed(const NumberedServerVector& servers, const DNSQuestion* dq)
+{
+  std::map<unsigned int, shared_ptr<DownstreamState>> circle = {};
+  unsigned int qhash = dq->qname->hash(g_hashperturb);
+  unsigned int sel = 0, max = 0;
+  shared_ptr<DownstreamState> ret = nullptr, last = nullptr;
+
+  for (const auto& d: servers) {
+    if (d.second->isUp()) {
+      // make sure hashes have been computed
+      if (d.second->hashes.empty()) {
+        d.second->hash();
+      }
+      {
+        ReadLock rl(&(d.second->d_lock));
+        const auto& server = d.second;
+        // we want to keep track of the last hash
+        if (max < *(server->hashes.rbegin())) {
```

Comment:
Don't we have an issue here if the weight is set to zero, or even a negative since it's a signed integer,?

Reply:
Didn't consider that as the weight value is bounded when the server is added (see https://github.com/PowerDNS/pdns/blob/04437f1a023bbe90f13e3f8372f1b49f3554c9bb/pdns/dnsdist-lua.cc#L269)

Reply:
Right, although it looks like we have a `Lua` binding for the `weight` member that does not check anything :'(

Reply:
Just added some checks around the `weight` member binding